### ISSUE TITLE
[RFC] [Sampling] Sampling with edge masks

### ISSUE THREAD

**Issue by BarclayII** [dmlc/dgl#4441]
## 🚀 Feature
Support sampling with a boolean edge mask indicating which edges can be included in message passing.

## Motivation
During validation, only the edges in the training set should be considered for message passing.  During testing, the edges in both the training set and the validation set should be considered for message passing.  If we have sampling methods that support edge masks, the user can just create a single graph with both the training and validation edges, and provide different masks to the sampler so that different subsets can be used during sampling for minibatch message passing.

## Alternatives
Construct two graphs, one with all the edges and the other with only the masked edges.  The problem with this approach is that it will consume twice as much space when the proportion of the masked edges is small, making it unsuitable for large-scale graphs.  Supporting two graphs in distributed training is also difficult since either the user will have to repartition the graph, or a significant refactor is necessary, because currently the distributed training design (load balancing, data partitioning, etc.) is centered around the assumption of having only a single graph.

## Pitch
Have an `include_mask: str` argument for the following methods:
* `subgraph`
* `edge_subgraph`
* `in_subgraph`
* `out_subgraph`
* `sample_neighbors`
* `sample_topk`

And maybe the following interfaces:
* `node_type_subgraph`
* `edge_type_subgraph`
* `random_walk`
* `node2vec`

The `include_mask` argument will be a string indicating the name of the boolean mask on the edges.  Only the edges with a True mask can appear in the results from the methods above.  Or in other words, the edges with a False mask will be skipped during sampling.

Example:
```python
g = dgl.graph(([0,1,2,3],[1,2,3,4]))
g.edata['mask'] = torch.BoolTensor([True, False, True, True])
sg = g.subgraph([0, 1, 2])
sg.edges()    # ([0, 1], [1, 2])
sg = g.subgraphs([0, 1, 2], include_mask='mask')
sg.edges()    # ([0, 1])
```

**Comment by mfbalin** [comment]
One can store the training and validation edges in a single graph with the constraint that all the training edges come before all the validation edges for each node's neighborlist. Then one can store an additional indptr called indptr2 to point to the start of the validation edges in each node's neighborlist. Then if one sets indptr_start=indptr & indptr_end=indptr2, then one only uses training edges. If one sets indptr_start=indptr & indptr_end=indptr+1, then one uses both training and validation edges. This requires the C++ interface to have additional methods such as get_indptr_start and get_indptr_end along with set_indptr_start and set_indptr_end to change the mode of the graph and the sampling methods to use the get_indptr_start and get_indptr_end methods to sample from the current mode.

**Comment by mfbalin** [comment]
The method described above can potentially also be used with directed graphs made into undirected, then one can store all the added edges after all the original edges, and a similar logic applies.

**Comment by BarclayII** [comment]
@mfbalin I guess your idea is to store two indptr arrays?  I feel it would complicate the implementation of the underlying CSR operators comparing to boolean masks.

**Comment by mfbalin** [comment]
Yes, the only thing that would change to iterate over the wanted set of edges is, instead of using indptr[u] and indptr[u+1] to iterate over u's neighbors, the user would iterate over indptr_start[u] and indptr_end[u]. One would need to do this only for the kernels where one wants to filter some edges, not touching other kernels would result into the original behaviour with indptr and indptr+1 to iterate over all the edges.

**Comment by kkranen** [comment]
> Yes, the only thing that would change to iterate over the wanted set of edges is, instead of using indptr[u] and indptr[u+1] to iterate over u's neighbors, the user would iterate over indptr_start[u] and indptr_end[u]. One would need to do this only for the kernels where one wants to filter some edges, not touching other kernels would result into the original behaviour with indptr and indptr+1 to iterate over all the edges.

Does this address any edge cases where Train/validation edges are not disjoint? This could cause a significant headache for multi-edges as well.

**Comment by BarclayII** [comment]
One issue I found with the current interface is that, the implementation based on this interface cannot introduce any additional data structure (such as pre-computing the number of nodes and the list of included neighbors).  The simplest implementation without additional data structure is reusing our non-uniform neighbor sampling algorithm, which will be a lot slower than neighbor sampling since it needs to compute the prefix sum every time for every node and do a binary search for every sampled neighbor.

**Comment by mfbalin** [comment]
> > Yes, the only thing that would change to iterate over the wanted set of edges is, instead of using indptr[u] and indptr[u+1] to iterate over u's neighbors, the user would iterate over indptr_start[u] and indptr_end[u]. One would need to do this only for the kernels where one wants to filter some edges, not touching other kernels would result into the original behaviour with indptr and indptr+1 to iterate over all the edges.
> 
> Does this address any edge cases where Train/validation edges are not disjoint? This could cause a significant headache for multi-edges as well.

If any edge is common between the training and the validation set, then those edges can be stored after edges that are training only and before edges that are validation only. Then one can store an indptr for training end, another indptr for validation start.

**Comment by BarclayII** [comment]
This is implemented in #4748 and #4749.
